core/src/main/scala/org/apache/spark/h2o/backends/internal/SpreadRDDBuilder.scala

```diff
@@ -18,106 +18,111 @@
+    } else if (nSparkExecAfter != nSparkExecBefore || nSparkExecAfter != currentWorkers) {
       // We detected change in number of executors
-      logInfo(s"Detected ${nSparkExecBefore} before, and ${nSparkExecAfter} spark executors after, backend#isReady=${isBackendReady()}! Retrying again...")
-      build(nretries - 1, 2*mfactor, 0)
+      logInfo(s"Detected $nSparkExecBefore before, and $nSparkExecAfter Spark executors after, backend#isReady=${isBackendReady()}! Retrying again...")
+      build(nretries - 1, 2 * mfactor, 0)
     } else if ((numTriesSame == conf.subseqTries)
-      || (numExecutors.isEmpty && numVisibleNodes == expectedWorkers)
-      || (numExecutors.isDefined && numExecutors.get == numVisibleNodes)) {
-      logInfo(s"Detected ${numVisibleNodes} spark executors for ${expectedWorkers} H2O workers!")
-      (new InvokeOnNodesRDD(visibleNodes, sc), visibleNodes)
+      || (numExecutors.isEmpty && currentWorkers == expectedWorkers)
+      || (numExecutors.isDefined && numExecutors.get == currentWorkers)) {
+      logInfo(s"Detected $currentWorkers spark executors for $expectedWorkers H2O workers!")
+      endpoints
     } else {
-      logInfo(s"Detected ${numVisibleNodes} spark executors for ${expectedWorkers} H2O workers, backend#isReady=${isBackendReady()}! Retrying again...")
-      build(nretries-1, mfactor, numTriesSame + 1)
+      logInfo(s"Detected $currentWorkers spark executors for $expectedWorkers H2O workers, backend#isReady=${isBackendReady()}! Retrying again...")
+      build(nretries - 1, mfactor, numTriesSame + 1)
     }
   }
 
-  /** Generates and distributes a flatfile around Spark cluster.
-    *
-    * @param distRDD simple RDD to fork execution on.
-    * @return list of node descriptions NodeDesc(executorId, executorHostName, -1)
-    */
-  def collectNodesInfo(distRDD: RDD[Int]): Array[NodeDesc] = {
-    // Collect flatfile - tuple of (executorId, IP, -1)
-    val nodes = distRDD.mapPartitionsWithIndex { (idx, it) =>
-      val env = SparkEnv.get
-      Iterator.single(NodeDesc(env.executorId, SharedBackendUtils.getHostname(env), -1))
-    }.collect()
-    // Take only unique executors
-    nodes.groupBy(_.nodeId).map(_._2.head).toArray.sortWith(_.nodeId < _.nodeId)
-  }
-
   /**
     * Return number of registered Spark executors
     */
-  private def numOfSparkExecutors = if (sc.isLocal) 1 else {
+  private def numOfSparkExecutors = if (isLocal) 1 else {
     val sb = sc.schedulerBackend
     sb match {
-      case b: LocalSchedulerBackend => 1
+      case _: LocalSchedulerBackend => 1
       case b: CoarseGrainedSchedulerBackend => b.getExecutorIds.length
       case _ => SparkEnv.get.blockManager.master.getStorageStatus.length - 1
     }
   }
 
   private def isBackendReady() = sc.schedulerBackend.isReady()
 }
+
+object RpcReferenceCache {
+  private var ref: RpcEndpointRef = _
+
+  def getRef(conf: SparkConf): RpcEndpointRef = synchronized {
```

Comment:
What object is used for locking in this case? Isn't  it ```this```?

Reply:
yup, this -> we should synchronize on `RpcReferenceCache`, thanks!

Reply:
I'm afraid that `RpcReferenceCache` is pointing to the same instance as `this`.  The reason of my question was that we should limit the scope of synchronization only to `RpcReferenceCache`. Another thread could acquire lock  outside of `RpcReferenceCache` and this might lead to problems. Apologize, I should have been more explicit about this problem.

This will cause a deadlock for instance: 
```
object SomeObject {
  assert(this == SomeObject)
  def someMethod() = synchronized {
    println(this)
    println("Hello World!")
  }
}

object Application extends App {
  SomeObject.synchronized{
    val thread = new Thread(
      new Runnable {
        def run(): Unit = {
          SomeObject.someMethod()
        }
      })
    thread.start()
    thread.join()
  }
}
```

I good practice is to create a private instance dedicated just for locking. WDYT?

Reply:
Great catch, thank you!